
\section{processes.xml}

\subsection{Rationale}

The process file is used to define cellular processes involved in
the production or degradation of macromolecules.
In proteins.xml, rnas.xml and dna.xml, we have described proteins in terms of
components (amino acid residues, nucteotides, protein cofactors).
processes.xml defines how components are assembled from metabolites and
what machines are needed to catalyze the assembly.

A typical example of a cellular process is protein translation.
Translation relies on ribosomes as a molecular machine,
and participates in protein production.
The metabolites used during proteins assembly include
charged and uncharged-tRNAs and GTP/GDP,
e.g.\ an alanine residue is obtained by using a charged tRNA-alanine and GTP as
reactants, generating an uncharged tRNA, GDP and water as a byproduct.

Theoretically, the translation of a protein could be written down as a single reaction,
with

with specific tRNAs used
to cite a few, which correspond to the processing costs of the ribosome.
Since the amino-acid composition of proteins is known from the genome sequence,
protein production could theoretically be described by metabolic-like reactions
involving charged-/uncharged-tRNAs, GTP/GDP, one per protein, according to the SBML format.
But specifying all these reactions explicitly for each protein species
would be a laborious and error-prone if we have to consider the entire cell.

Indeed, macromolecules such as proteins typically undergo several processes
(translation, chaperoning, translocation),
making them effectively processed by several machines operating in series.
To account for such a level of detail,
one would need to define one reaction per macromolecule per process undergone,
with the appropriate molecular machine such as an enzyme or a ribosome.
The description of models in SBML is protein- or gene- centred,
meaning that the fate of each protein has to be described explicitly by individual reactions.
Therefore, if a resource allocation model contains 1000 proteins,
and if all proteins require chaperones for folding,
1000 reactions for protein synthesis and 1000 reactions for chaperoning,
2000 reactions would have to be listed in the SBML file.
The SBML file would thus grow very rapidly and would contain thousands of closely related reactions,
making it extremely difficult to check the consistency of reactions
or to update them upon addition of a new process.
This problem would become even more severe if of processes are further subdivided
into smaller steps, e.g., ribosome binding, chain elongation, and ribosome unbinding.

An alternative description consists in defining only one template reaction per cellular process,
which produces a list of macromolecules.
This definition is now process-centred instead of being protein-centred,
and offers a simpler and clearer view of how macromolecules are produced.
This process-centred representation is generic,
flexible and adaptable to a large set of bacterial cells.
It was also shown to be an efficient framework to interrelate various cellular entities
(metabolites, proteins, cellular processes, sequences, etc.)
with their physical or biochemical properties (e.g. kinetic parameters, etc.)
within whole-cell models.
Unfortunately, such template reactions cannot be encoded with current SBML releases.
Consequently, for encoding resource allocation models in RBApy,
we chose not to follow a protein-centred representation as in SBML,
but the formal process-centred representation of the bioontology BiPON.
The principle behind BiPON is to break down the cell into intertwined cellular processes,
where each cellular process is described as an input/output system.
In RBApy, every cellular process is described by (1) the molecular machine that
catalyzes the process, (2) the list of macromolecules to be processed or produced,
(3) the processing costs, and (4) the efficiency of the molecular machine in catalyzing the process,
which corresponds to the rate of the process per amount of process machinery
(e.g. translation rate for ribosomes).
We designed an XML template that helps to define the processing costs,
the composition of the molecular machine and the list of macromolecules to be processed
(see Supplementary Note S5).
Then, defining a new cellular process is straightforward and necessitates
only a few steps for the user (see Supplementary Note S5).
The cost of all newly targeted processed macromolecules is updated automatically.
For example, if a model needsis supposed to cover protein secretion,
translocation processes may be added.
As soon as a protein is listed as an input of one of the translocation processes,
its overall production cost will be updated (Supplementary Figure S11).
Molecular machines can be easily described by listing their individual components
(proteins, RNAs, metabolites) together with their stoichiometries.


\process{}es can be seen as template reactions that specify how
\macromolecule{}s are produced and degraded based on their \component{}s.
A typical example is translation: a \machinery{} (the ribosome) assembles
proteins based on their amino acid sequence (\component{}s).
Every amino acid is assembled using metabolic \species, in particular tRNAs.
However, the \species{} that are used differ from amino acid to amino acid.
The \processingmap{} explicits how every \component{} is processed, i.e.
what \species{} are consumed (e.g. charged tRNAs, GTP) and what \species{} are
produced (e.g. uncharged tRNAs, GDP).
Briefly put, \process{}es define one reaction per \component(), frow which
the production/degradation reactions of all \macromolecule{}s can be deduced.

\macromolecule{}s can go through several \process{}es:
for example, proteins may undergo translation, folding and translocation.
The \processings{} element is used to specify what macromolecule goes
through which process.
The overall production/degradation reaction of a \macromolecule{} reflects
all the \process{}es it traverses.

\subsection{RBAProcesses}
\label{sec:rba_processes}

The outermost part of the process file is an instance of class
\rbaprocesses, shown in Figure~\ref{fig:processes_doc}.

\begin{figure}
  \centering
  \includegraphics[scale=0.8]{figures/processes_doc}
  \caption{XML structure of process document.}
\label{fig:processes_doc}
\end{figure}

\rbaprocesses{} has no simple attributes.
It contains exactly one instance of \textbf{ListOfProcesses}
and \textbf{ListOfProcessingMaps}.


\subsection{Process}
\label{sec:process}

The \process{} class is used to define cellular processes
(Fig.~\ref{fig:processes_process}).

\begin{figure}
  \centering
  \includegraphics[scale=0.8]{figures/processes_process}
  \caption{Class used to store processes.}
\label{fig:processes_process}
\end{figure}

A \process{} revolves around 2 optional substructures.
The \machinery{} is the molecular entity enabling the process
(\textit{e.g.} ribosome for translation.)
Each machinery unit has a limited production/degradation capacity.
Every macromolecule produced by a process has a metabolite cost
(metabolites needed to produce/degrade it and byproducts).
However, if a machinery is defined, there is an additional cost
to produce the machinery that will enable the production/degradation of the
target.
This is similar to the production of \enzyme{}s in order to catalyze
metabolic \reaction{}s.

\processings{} define the sets of macromolecules that a process
produces or degrades.
The production reaction of a \macromolecule{} is determined by the \process{}es
it goes through.
For example, a protein's production reaction is defined by listing
the protein as an input in the \processings{} of the translation process.
If a protein is not listed as an input of any process, its production reaction
is empty, meaning that it does not cost anything to produce the protein.
\processings{} break down \macromolecule{}s in metabolic \species{}
and \machinery{} costs.

\paragraph{The \textit{id} attribute}
The \textbf{id} attribute is a string defining the identifier of a process.

\paragraph{The \textit{name} attribute}
The \textbf{name} attribute is a string that can be used to give the process
a more human understandable name.


\subsection{Machinery}
\label{sec:machinery}

The \machinery{} class defines the machinery used by a process
(Fig.~\ref{fig:processes_process}).
\machinery{} has no simple attributes.
If a \machinery{} is defined, it defines a \emph{capacity constraint}.
Every \machinery{} unit is produced accordint to the reaction
defined by a \machinerycomposition.
Every unit also has a capacity defined by a \targetvalue.
The capacity defines how many targets a \machinery{} can process in 1 unit of
time.
Total capacity (base capacity multiplied by number of \machinery{} units)
must always exceed the number of targets produced.


\subsection{MachineryComposition}
\label{sec:machinery_composition}

The \machinerycomposition{} class defines the assembly of a complex
molecular machinery (Fig.~\ref{fig:processes_process}).
\machinerycomposition{} has no simple attributes.
It contains two \textbf{ListOfSpeciesReferences}.
One is for reactants, the other for potential byproducts of the assembly reaction
of the complex itself (e.g. GDP when connecting ribosomal subunits).
Note that in this case, \speciesreference{}s can refer to \emph{both}
metabolic \species{} and \macromolecule{}s.
The assembly reaction should contain obvious components of the machinery,
but also metabolic costs related to assembly (such as ATP/GTP costs)
\emph{unless} these costs are already covered by a process.


\subsection{Processings}
\label{sec:processings}

The \processings{} relates \macromolecule{}s production/degradation to
some given \process{}es (Fig.~\ref{fig:processes_process}).
\processings{} has no simple attributes.
It may contain two \textbf{ListOfProcessings}, one for production and one for
degradation.


\subsection{Processing}
\label{sec:processing}

The \processing{} class defines how \macromolecule{}s are produced/degraded
(Fig.~\ref{fig:processes_process}).
\processing{} is used to break down \macromolecule{}s into metabolites
by linking them to a \processingmap.
It contains one \textbf{ListOfSpeciesReferences} that lists \macromolecule{}s
that are inputs of this process.
In this context, the species of a \speciesreference{} must be a macromolecule
and the stoichiometry attribute is ignored.

\paragraph{The \textit{processingMap} attribute}
The \textbf{processingMap} attribute must match the identifier of a
\processingmap.
This \processingmap{} will be used to compute the production/degradation
reaction of \macromolecule{}s, as well as \machinery{} costs.

\paragraph{The \textit{set} attribute}
The \textbf{set} attribute must refer to a \macromolecule{} set.
Currently, the only acceptable values are \textbf{protein}, \textbf{rna}
and \textbf{dna}.
\macromolecule{}s that are listed as input must belong to this set.


\subsection{ProcessingMap}
\label{sec:processing_map}

The \processingmap{} class is used to convert \macromolecule{}s in
metabolic and machinery costs (Fig.\ref{fig:processes_processing_map}).

\begin{figure}
  \centering
  \includegraphics[scale=0.8]{figures/processes_processing_map}
  \caption{Class used to compute production/degradation of macromolecules.}
\label{fig:processes_processing_map}
\end{figure}

There are two types of processings.
The \constantprocessing{} lists metabolites that are always consumed or
produced when processing a macromolecule, no matter its composition
(\textit{e.g.} translation initiation).
The \textbf{ListOfComponentProcessing} container details
\componentprocessing{}s depending on the individual
\component{}s of the \macromolecule{}.
They cover metabolites used to assemble the \component{} onto the nascent
\macromolecule{}.
They also cover machinery costs, \textit{i.e.} how many \machinery{} units
are needed to assemble the \component{}.

\paragraph{The \textit{id} attribute}
The \textbf{id} attribute is a string defining the identifier of a
processing map.


\subsection{ConstantProcessing}
\label{sec:constant_processing}

The \constantprocessing{} class defines metabolites consumed and byproducts
generated by an assembly process (Fig.\ref{fig:processes_processing_map}).
It contains two \textbf{ListOfSpeciesReferences}, one for metabolites
consumed and one for metabolites produced.
Note that in this context, a \speciesreference{} must refer to a
metabolic \species.


\subsection{ComponentProcessing}
\label{sec:component_processing}

The \componentprocessing{} class defines metabolites consumed and byproducts
generated when assembling a specific \component{}
(Fig.\ref{fig:processes_processing_map}).
It contains two \textbf{ListOfSpeciesReferences}, one for metabolites
consumed and one for metabolites produced.
Note that in this context, a \speciesreference{} must refer to a
metabolic \species.
Additionally, it defines a machinery cost used in a \machinery{}'s
capacity constraint.

\paragraph{The \textit{component} attribute}
The \textbf{component} attribute is a string that must match the identifier
of a \component{}.

\paragraph{The \textit{machineryCost} attribute}
The \textbf{machineryCost} attribute is a real value that is used to
compute how many \machinery{} units are needed to assemble the \component{}.
For example, let the machinery cost for the processing of an amino acid be 1.
The capacity of the \machinery{} (the ribosome) is the number of amino acids
it can assemble per unit of time.
The machinery cost allows to compute how many ribosomes are needed
to produce the \component{} and, in the end, the \macromolecule{}
(in this example the number of amino acids divided by the ribosome's capacity).
